From 6222c14ca051fe15a9a95e86d8f9cc6b7880cdc4 Mon Sep 17 00:00:00 2001
From: Michael Pauzner <mpauzner@habana.ai>
Date: Tue, 29 Apr 2025 13:26:43 +0300
Subject: [PATCH] [SW-221409] gaudi3: serdes: create backup serdes image

Gaudi3 firmware, prior to v1.22, has hardcodded serdes table.
Starting v1.22, boot-fit will read serdes configuration from serdes
lfs partition in flash. However, during transition period,
the existing flash image may not contain serdes partition.

Thus, create backup serdes configuration for boot-fit.
The image will be loaded to an unused RAM of die1 arc.

Note: due of limitation of gaudi3 preboot, this feature
will work with prebbot v1.21 and above.


in die1 unused RAM.

Change-Id: I099ec005eb5cfde457531dcfec6b2b98caa7aba5
Signed-off-by: Michael Pauzner <mpauzner@habana.ai>
---

diff --git a/board/habanalabs/gaudi3/post-image-fit.sh b/board/habanalabs/gaudi3/post-image-fit.sh
index a54c8f8..77c7706 100644
--- a/board/habanalabs/gaudi3/post-image-fit.sh
+++ b/board/habanalabs/gaudi3/post-image-fit.sh
@@ -107,3 +107,7 @@
 ITS_SED_PARAMS+="s/@MGMT_STARTADDR_LO@/${MGMT_STARTADDR_LO}/g;"
 ITS_SED_PARAMS+="s/@EVENT_STARTADDR_HI@/${ARC_EVENT_ICCM_HI}/g;"
 ITS_SED_PARAMS+="s/@EVENT_STARTADDR_LO@/${EVENT_STARTADDR_LO}/g;"
+
+# serdes backup in Die1, Arc1 DCCM
+ITS_SED_PARAMS+="s/@D1_ARC1_DCCM_HI@/${D1_ARC1_DCCM_HI}/g;"
+ITS_SED_PARAMS+="s/@D1_ARC1_DCCM_LO@/${D1_ARC1_DCCM_LO}/g;"
diff --git a/board/habanalabs/gaudi3/zephyr.its b/board/habanalabs/gaudi3/zephyr.its
index 4850090..8b69bd7 100644
--- a/board/habanalabs/gaudi3/zephyr.its
+++ b/board/habanalabs/gaudi3/zephyr.its
@@ -125,6 +125,18 @@
 				algo = "@SHA_VERSION@";
 			};
 		};
+
+		serdes_backup_lfs {
+			description = "SerDes configuration in littlefs binary image";
+			data = /incbin/("gaudi3-serdes-config.bin");
+			type = "standalone";
+			core-id = <@FIT_CORE_ID_D1_PSOC_ARC1@>;
+			compression = "none";
+			loadaddr = <@D1_ARC1_DCCM_HI@ @D1_ARC1_DCCM_LO@>;
+			hash {
+				algo = "@SHA_VERSION@";
+			};
+		};
 	};
 
 	configurations {
@@ -135,7 +147,8 @@
 			description = "Habanalabs device preboot";
 			loadables = "zephyr-pid-iccm", "zephyr-pid-dccm", "zephyr-mgmt-iccm",
 				    "zephyr-mgmt-dccm", "zephyr-mgmt-reloc", "zephyr-mgmt-offload",
-				    "zephyr-event-iccm", "zephyr-event-dccm", "external_fw";
+				    "zephyr-event-iccm", "zephyr-event-dccm", "external_fw",
+				    "serdes_backup_lfs";
 		};
 	};
 };
